launchable/test_runners/cts.py

```diff
@@ -0,0 +1,178 @@
+            </StackTrace>
+          </Failure>
+        </Test>
+      </TestCase>
+    </Module>
+    """
+    class TestResult:
+        def __init__(self, test_case_name: str, test_name: str, result: str, stdout: str, stderr: str):
+            self.test_case_name = test_case_name
+            self.test_name = test_name
+            self.result = result
+            self.stdout = stdout
+            self.stderr = stderr
+
+        def case_event_status(self):
+            if self.result == "fail":
+                return CaseEvent.TEST_FAILED
+            elif self.result == "ASSUMPTION_FAILURE" or self.result == "IGNORED":
+                return CaseEvent.TEST_SKIPPED
+
+            return CaseEvent.TEST_PASSED
+
+    def build_record_test_path(module: str, test_case: str, test: str):
+        return [
+            {"type": "Module", "name": module},
+            {"type": "TestCase", "name": test_case},
+            {"type": "Test", "name": test},
+        ]
+
+    events = []
+    tree = ET.parse(p)
+
+    for module in tree.iter('Module'):
+        test_results = []
+        total_duration = module.get("runtime", "0")
+        module_name = module.get("name", "")
+
+        for test_case in module.iter("TestCase"):
+            test_case_name = test_case.get('name', "")
+
+            for test in test_case.iter("Test"):
+                result = test.get('result', "")
+                test_name = test.get("name", "")
+                stdout = ""
+                stderr = ""
+
+                failure = test.find('Failure')
+                if failure:
+                    stack_trace = ""
+                    stack_trace_element = failure.find("StackTrace")
+                    if stack_trace_element is not None:
+                        stack_trace = str(stack_trace_element.text)
+
+                    stdout = failure.get("message", "")
+                    stderr = stack_trace
+
+                test_results.append(TestResult(
+                    test_case_name=test_case_name,
+                    test_name=test_name,
+                    result=result,
+                    stdout=stdout,
+                    stderr=stderr))
+
+        if len(test_results) == 0:
+            continue
+
+        test_duration_msec_per_test = int(total_duration) / len(test_results)
+
+        for test_result in test_results:  # type: TestResult
+            if module_name == "" or test_result.test_case_name == "" or test_result.test_name == "":
+                continue
+
+            events.append(CaseEvent.create(
+                test_path=build_record_test_path(module_name, test_result.test_case_name, test_result.test_name),
+                duration_secs=float(test_duration_msec_per_test / 1000),
+                status=test_result.case_event_status(),
+                stdout=test_result.stdout,
+                stderr=test_result.stderr,
+            ))
+
+    return (x for x in events)
```

Comment:
[question]
Is this intended to return Generator?

Reply:
Yes, it is.
I aligned to nunit and raw profile.
https://github.com/launchableinc/cli/blob/main/launchable/test_runners/raw.py#L149
https://github.com/launchableinc/cli/blob/main/launchable/test_runners/nunit.py#L103

Also, this cts report files may include too many test cases, so I adapted memory-efficient approch

Reply:
ok